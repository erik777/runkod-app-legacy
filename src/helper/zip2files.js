import {DISABLED_FILES} from '../constants';

const JSZip = require('jszip');

const fileName = (path) => {
  const sp = path.split('/');

  return sp[sp.length - 1];
};

const fileType = (fileName) => {
  const sf = fileName.split('.');
  const ext = sf[sf.length - 1].toLowerCase();
  switch (ext) {
    case 'html':
      return 'text/html';
    case 'css':
      return 'text/css';
    case 'png':
      return 'image/png';
    case 'js':
      return 'application/javascript';
    default:
      return '';
  }
};

const extractFile = async (z, root) => {
  const buffer = await z.async('arraybuffer');

  const name = fileName(z.name);
  const path = z.name.split('/').filter(x => x !== root).filter(x => x !== name).filter((x) => x !== '').join('/');
  const type = fileType(name);

  return {path, name, type, size: buffer.byteLength, buffer}
};


export default async (buffer) => {
  const zip = await JSZip.loadAsync(buffer);

  const root = zip.filter((relativePath, file) => file.dir)[0].name.replace('/', '');

  const zipFiles = zip.filter(function (relativePath, file) {
    if (file.dir) {
      return false;
    }

    const fn = fileName(relativePath);

    return !DISABLED_FILES.includes(fn);
  });

  return await Promise.all(zipFiles.map(z => extractFile(z, root)));
}
