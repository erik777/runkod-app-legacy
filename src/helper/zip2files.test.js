import zip2files from './zip2files.js';

const fs = require('fs');

it('1- zip2files',async () => {
  const input = fs.readFileSync('./public/files/runkod-placeholder-dist.zip', ).buffer;
  const result = await zip2files(input);
  expect(result).toMatchSnapshot();
});
