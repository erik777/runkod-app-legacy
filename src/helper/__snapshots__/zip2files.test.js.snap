// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1- zip2files 1`] = `
Array [
  Object {
    "buffer": ArrayBuffer [],
    "name": "index.html",
    "path": "",
    "size": 290,
    "type": "text/html",
  },
  Object {
    "buffer": ArrayBuffer [],
    "name": "main.js",
    "path": "",
    "size": 70,
    "type": "application/javascript",
  },
  Object {
    "buffer": ArrayBuffer [],
    "name": "logo.png",
    "path": "",
    "size": 4542,
    "type": "image/png",
  },
  Object {
    "buffer": ArrayBuffer [],
    "name": "style.css",
    "path": "",
    "size": 654,
    "type": "text/css",
  },
]
`;
